imap/imap.go

```diff
@@ -0,0 +1,353 @@
+func Validate(s *models.IMAP) error {
+
+	err := s.Validate()
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	s.Host = s.Host + ":" + strconv.Itoa(int(s.Port)) // Append port
+	mailServer := Mailbox{
+		Host:   s.Host,
+		TLS:    s.TLS,
+		User:   s.Username,
+		Pwd:    s.Password,
+		Folder: s.Folder}
+
+	client, err := mailServer.newClient()
+	if err != nil {
+		log.Error(err.Error())
+	} else {
+		client.Close(true)
+		client.Logout(30 * time.Second)
+	}
+	return err
+}
+
+// Response is a helper struct to wrap the email responses and possible errors.
+type Response struct {
+	Email Email
+	Err   error
+}
+
+// newClient will initiate a new IMAP connection with the given creds.
+func (mbox *Mailbox) newClient() (*imap.Client, error) {
+	var client *imap.Client
+	var err error
+	if mbox.TLS {
+		client, err = imap.DialTLS(mbox.Host, new(tls.Config))
+		if err != nil {
+			return client, err
+		}
+	} else {
+		client, err = imap.Dial(mbox.Host)
+		if err != nil {
+			return client, err
+		}
+	}
+
+	_, err = client.Login(mbox.User, mbox.Pwd)
+	if err != nil {
+		return client, err
+	}
+
+	_, err = imap.Wait(client.Select(mbox.Folder, mbox.ReadOnly))
+	if err != nil {
+		return client, err
+	}
+
+	return client, nil
+}
+
+const dateFormat = "02-Jan-2006"
+
+// findEmails will run a find the UIDs of any emails that match the search.:
+func findEmails(client Client, search string, since *time.Time) (*imap.Command, error) {
+	var specs []imap.Field
+	if len(search) > 0 {
+		specs = append(specs, search)
+	}
+
+	if since != nil {
+		sinceStr := since.Format(dateFormat)
+		specs = append(specs, "SINCE", sinceStr)
+	}
+
+	// get headers and UID for UnSeen message in src inbox...
+	cmd, err := imap.Wait(client.UIDSearch(specs...))
+	if err != nil {
+		return &imap.Command{}, fmt.Errorf("uid search failed: %s", err)
+	}
+	return cmd, nil
+}
+
+const GenerateBufferSize = 100
+
+func (mbox *Mailbox) generateMail(search string, since *time.Time, markAsRead, delete bool) (chan Response, error) {
+	responses := make(chan Response, GenerateBufferSize)
+	client, err := mbox.newClient()
+	if err != nil {
+		close(responses)
+		return responses, fmt.Errorf("failed to create IMAP connection: %s", err)
+	}
+
+	go func() {
+		defer func() {
+			client.Close(true)
+			client.Logout(30 * time.Second)
+			close(responses)
+		}()
+
+		var cmd *imap.Command
+		// find all the UIDs
+		cmd, err = findEmails(client, search, since)
+		if err != nil {
+			responses <- Response{Err: err}
+			return
+		}
+		// gotta fetch 'em all
+		getEmails(client, cmd, markAsRead, delete, responses)
+	}()
+
+	return responses, nil
+}
+
+func getEmails(client Client, cmd *imap.Command, markAsRead, delete bool, responses chan Response) {
+	seq := &imap.SeqSet{}
+	msgCount := 0
+	for _, rsp := range cmd.Data {
+		for _, uid := range rsp.SearchResults() {
+			msgCount++
+			seq.AddNum(uid)
+		}
+	}
+
+	// nothing to request?! why you even callin me, foolio?
```

Comment:
While this made me laugh, let's remove this for now to keep things as clean as possible 😆

Reply:
Done 🙃